File: src/Message.ts
// WebSocketRT: Real-time websocket library
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

import { MessageCallback, MessageCallbackEvents, MessageCallbackHandler } from './MessageCallbackHandler';

/** Holds one complete message. Messages are broken up into one or more frames while in transport. */
export class Message {
  /**
   * Constructor (Unlike the C# implementation, which has two overloaded constructors, TypeScript uses a union type to
   * implement both as a single constructor)
   * @param payloadLength Optional payload length. If set, a buffer will be automatically created for `payload`.
   *    For outgoing messages, the caller may also pass in a `Uint8Array` containing payload data.
   * @param outgoing `true` for outgoing messages; `false` for incoming
   */
  public constructor(payload?: Uint8Array | number, outgoing = true) {
    this.outgoing = outgoing;

    if (typeof payload === 'number' && payload > 0) {
      this._payload = new Uint8Array(payload);
    } else if (typeof payload === 'object') {
      if (!outgoing) {
        // Payload may only be specified when creating outgoing messages
        throw new Error('Payload on incoming');
      }
      this._payload = payload;
    }
  }

  /** `true` for outgoing messages; `false` for incoming */
  public readonly outgoing: boolean;

  /** Optional header (64 bytes max) */
  public getHeader(): Uint8Array {
    return this._header;
  }

  /**
   * Internal. Do not call outside the transport layer. Unfortunately set has to be public, as TypeScript doesn't have
   * the equivalent of C#'s internal scope.
   */
  public _setHeader(value: Uint8Array): void {
    if (this.outgoing) {
      // Cannot set Header on outgoing Message
      throw new Error('Header on outgoing');
    }
    this._header = value;
  }
  private _header: Uint8Array;

  /**
   * Payload data. Note that the size of the array is the expected length, not the actual length received.
   * Always check `getBytesReceived()` for the actual number received so far, and don't read past that
   * point in this array.
   */
  public getPayload(): Uint8Array {
    return this._payload;
  }
  private _payload: Uint8Array;

  /** Number of payload bytes received */
  public getBytesReceived(): number {
    return this.outgoing ? this._payload.length : this._bytesReceived;
  }

  /** Internal. Do not access outside the transport layer. */
  public _bytesReceived = 0;

  /** `true` if the payload has been fully received; `false` otherwise */
  public isComplete(): boolean {
    return this._payload.length === this.getBytesReceived();
  }

  /** `true` if the message has been cancelled and will never complete */
  public isCancelled(): boolean {
    return this._isCancelled;
  }

  /** Internal. Do not set outside the transport layer. */
  public _isCancelled = false;

  /** Reads the payload property as a JSON object */
  public getPayloadAsJson(): any {
    // Unpack the server's message
    const payloadString: string = String.fromCharCode.apply(null, this._payload);
    return JSON.parse(payloadString);
  }

  /** Writes the JSON notation for an object into the payload property */
  public setPayloadAsJson(obj: any): void {
    const payloadString = JSON.stringify(obj);

    // TODO: This part can probably be optimized. See: http://code.google.com/p/stringencoding/
    const payloadLength = payloadString.length;
    this._payload = new Uint8Array(payloadString.length);
    for (let n = 0; n < payloadLength; n++) {
      this._payload[n] = payloadString.charCodeAt(n);
    }
  }

  /**
   * Registers a callback to be executed on message events
   * @param callback Callback function
   * @param events Events that trigger the callback
   */
  public registerCallback(callback: MessageCallback, events = MessageCallbackEvents.Complete): void {
    if (this.outgoing) {
      // Cannot register callbacks on outgoing messages
      throw new Error('Callback on outgoing');
    }

    if (events === 0) {
      // Event mask is required
      throw new Error('Event mask');
    }

    if ((events & MessageCallbackEvents.NewMessage) !== 0) {
      // Cannot register for the NewMessage event at the message level
      throw new Error('New at message');
    }

    this._callbacks.registerCallback(callback, events);
  }

  /**
   * Internal. Invoked from the dispatch loop to execute registered callbacks.
   * @param callbacks Collection of registered callbacks. If `null`, the callbacks registered to this particular message
   *    are executed.
   * @returns Number of callback functions executed
   */
  public _executeCallbacks(callbacks: MessageCallbackHandler = null): number {
    if (this.outgoing) {
      throw new Error('Outgoing assert');
    }

    // Compute which events to send
    let events = MessageCallbackEvents.PayloadReceived;
    if (!this._sentNewMessageCallback) {
      if (callbacks) {
        // This is hacky, but we get called twice, once for message-level callbacks where
        // callbacks == null, followed by a second time for connection-level callbacks where
        // callbacks != null. To ensure we deliver to both, we only set the boolean on the second call to
        // this function. There are no race conditions, since the dispatch loop is single-threaded.
        this._sentNewMessageCallback = true;
      }
      events |= MessageCallbackEvents.NewMessage;
    }
    if (!this._sentCompleteCallback && this.isComplete()) {
      if (callbacks) {
        // Same hack as above, by for Complete events
        this._sentCompleteCallback = true;
      }
      events |= MessageCallbackEvents.Complete;
    }
    if (this.isCancelled()) {
      if ((events & MessageCallbackEvents.NewMessage) !== 0) {
        // If we never delivered the NewMessage event, and the message is already cancelled, let's just not
        // tell anyone...
        return 0;
      }

      // Likewise, no reason to deliver a PayloadReceived event if the message is never going to complete...
      events = MessageCallbackEvents.Cancelled;
    }

    // Send the callbacks
    if (!callbacks) {
      return this._callbacks.executeCallbacks(this, events);
    } else {
      return callbacks.executeCallbacks(this, events);
    }
  }

  private _callbacks = new MessageCallbackHandler();

  // The two booleans below ensure we only ever deliver one set of NewMessage and one set of Complete callbacks
  private _sentNewMessageCallback = false;
  private _sentCompleteCallback = false;
}

File: dotnet/WebSocketRT/Message.cs
// WebSocketRT: Real-time websocket library
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

using System;
using System.Diagnostics;
using System.Text;
using System.Text.Json;

namespace LeoSingleton.WebSocketRT
{
    /// <summary>
    /// Holds one complete message. Messages are broken up into one or more frames while in transport.
    /// </summary>
    public class Message
    {
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="payloadLength">
        /// Optional payload length. If set, a buffer will be automatically created for <see cref="Payload"/>.
        /// </param>
        /// <param name="outgoing">True for outgoing messages; false for incoming</param>
        public Message(int payloadLength, bool outgoing = true)
        {
            Outgoing = outgoing;

            if (payloadLength > 0)
            {
                Payload = new byte[payloadLength];
            }
        }

        /// <summary>
        /// Constructor for outgoing messages only
        /// </summary>
        /// <param name="payload">Buffer for message payload</param>
        public Message(byte[] payload)
        {
            Outgoing = true;
            Payload = payload;
        }

        /// <summary>
        /// True for outgoing messages; false for incoming
        /// </summary>
        public bool Outgoing { get; private set; }

        /// <summary>
        /// Optional header (64 bytes max). This field is only used to return the header on incoming messages and not
        /// set on outgoing.
        /// </summary>
        public byte[] Header
        {
            get { return _Header; }
            internal set
            {
                if (Outgoing)
                {
                    throw new InvalidOperationException("Cannot set Header on outgoing Message");
                }
                _Header = value;
            }
        }
        private byte[] _Header;

        /// <summary>
        /// Payload data. Note that the size of the array is the expected length, not the actual length received.
        /// Always check <see cref="BytesReceived"/> for the actual number received so far, and don't read past that
        /// point in this array.
        /// </summary>
        public byte[] Payload { get; private set; }

        /// <summary>
        /// Number of payload bytes received
        /// </summary>
        public int BytesReceived
        {
            get
            {
                // Outgoing messages are always fully received. Only use _BytesReceived for incoming.
                return Outgoing ? Payload.Length : _BytesReceived;
            }

            set
            {
                if (Outgoing)
                {
                    throw new InvalidOperationException("Cannot set BytesReceived on outgoing Message");
                }

                _BytesReceived = value;
            }
        }

        private int _BytesReceived;

        /// <summary>
        /// True if the payload has been fully received; false otherwise
        /// </summary>
        public bool IsComplete => Payload.Length == BytesReceived;

        /// <summary>
        /// True if the message has been cancelled and will never complete
        /// </summary>
        public bool IsCancelled
        {
            get { return _IsCancelled; }
        }
        internal bool _IsCancelled;

        /// <summary>
        /// Reads the payload property as a JSON object
        /// </summary>
        /// <typeparam name="T">Expected object type of the payload</typeparam>
        /// <returns>Payload decoded from a JSON object</returns>
        public T GetPayloadAsJson<T>()
        {
            var payloadString = Encoding.UTF8.GetString(Payload);
            return JsonSerializer.Deserialize<T>(payloadString);
        }

        /// <summary>
        /// Writes the JSON notation for an object into the payload property
        /// </summary>
        /// <param name="obj">Payload to be encoded as JSON</param>
        public void SetPayloadAsJson<T>(T obj)
        {
            var payloadString = JsonSerializer.Serialize(obj);
            Payload = Encoding.UTF8.GetBytes(payloadString);
        }

        /// <summary>
        /// Registers a callback to be executed on message events
        /// </summary>
        /// <param name="callback">Callback function</param>
        /// <param name="events">Events that trigger the callback</param>
        public void RegisterCallback(MessageCallback callback,
            MessageCallbackEvents events = MessageCallbackEvents.Complete)
        {
            if (Outgoing)
            {
                throw new InvalidOperationException("Cannot register callbacks on outgoing messages");
            }

            if (events == 0)
            {
                throw new ArgumentException("Event mask is required", nameof(events));
            }

            if ((events & MessageCallbackEvents.NewMessage) != 0)
            {
                throw new ArgumentException("Cannot register for the NewMessage event at the message level",
                    nameof(events));
            }

            _Callbacks.RegisterCallback(callback, events);
        }

        /// <summary>
        /// Invoked from the dispatch loop to execute registered callbacks
        /// </summary>
        /// <param name="callbacks">
        /// Collection of registered callbacks. If null, the callbacks registered to this particular message are
        /// executed.
        /// </param>
        /// <returns>Number of callback functions executed</returns>
        internal int ExecuteCallbacks(MessageCallbackHandler callbacks = null)
        {
            Debug.Assert(!Outgoing);

            // Compute which events to send
            MessageCallbackEvents events = MessageCallbackEvents.PayloadReceived;
            if (!_SentNewMessageCallback)
            {
                if (callbacks != null)
                {
                    // This is hacky, but we get called twice, once for message-level callbacks where
                    // callbacks == null, followed by a second time for connection-level callbacks where
                    // callbacks != null. To ensure we deliver to both, we only set the boolean on the second call to
                    // this function. There are no race conditions, since the dispatch loop is single-threaded.
                    _SentNewMessageCallback = true;
                }
                events |= MessageCallbackEvents.NewMessage;
            }
            if (!_SentCompleteCallback && IsComplete)
            {
                if (callbacks != null)
                {
                    // Same hack as above, by for Complete events
                    _SentCompleteCallback = true;
                }
                events |= MessageCallbackEvents.Complete;
            }
            if (IsCancelled)
            {
                if (events.HasFlag(MessageCallbackEvents.NewMessage))
                {
                    // If we never delivered the NewMessage event, and the message is already cancelled, let's just not
                    // tell anyone...
                    return 0;
                }

                // Likewise, no reason to deliver a PayloadReceived event if the message is never going to complete...
                events = MessageCallbackEvents.Cancelled;
            }

            // Send the callbacks
            if (callbacks == null)
            {
                return _Callbacks.ExecuteCallbacks(this, events);
            }
            else
            {
                return  callbacks.ExecuteCallbacks(this, events);
            }
        }

        private MessageCallbackHandler _Callbacks = new MessageCallbackHandler();

        // The two booleans below ensure we only ever deliver one set of NewMessage and one set of Complete callbacks
        private bool _SentNewMessageCallback = false;
        private bool _SentCompleteCallback = false;
    }
}
